### PR TITLE
fix: Align error returned when a grant_type was requested that's not allowed for a client.

### PR DESCRIPTION
**Pull request by maennel**
## Related issue

None.

## Proposed changes

This fixes a minor difference between Ory Fosite and the OAuth2.0 specification on errors to be returned.

According to https://tools.ietf.org/html/rfc6749#section-5.2 the error to be returned if a client requests a grant_type it's not allowed to is 'unauthorized_client'.
The current implementation returns 'invalid_grant'.

This PR changes that behaviour and also adds or adapts tests where needed.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have read the [security policy](../security/policy)
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation within the code base (if appropriate)

## Further comments

No.

### CODE REVIEW

File: authorize_error_test.go [hunk 1]
```diff
@@ -70,12 +70,12 @@ func TestWriteAuthorizeError(t *testing.T) {
...
 			mock: func(rw *MockResponseWriter, req *MockAuthorizeRequester) {
 				req.EXPECT().IsRedirectURIValid().Return(false)
-				rw.EXPECT().Header().Return(header)
+				rw.EXPECT().Header().AnyTimes().Return(header)
 				rw.EXPECT().WriteHeader(http.StatusBadRequest)
 				rw.EXPECT().Write(gomock.Any())
 			},
 			checkHeader: func(t *testing.T, k int) {
-				assert.Equal(t, "application/json", header.Get("Content-Type"))
+				assert.Equal(t, "application/json;charset=UTF-8", header.Get("Content-Type"))
```
Comment (maennel) [inline authorize_error_test.go:78]: Note that these changes were needed to have tests passing. They are not semantically related to the rest of this PR.

Comment (mitar) [reply authorize_error_test.go:78]: I think these changes are already done in https://github.com/ory/fosite/pull/466/files

I suggest you remove them from this PR for now.

Comment (maennel) [reply authorize_error_test.go]: Done.

**Review by aeneasr — COMMENTED** [review]
Thank you, could you please fix the test? :)

**Comment by mitar** [comment]
I think you should first merge https://github.com/ory/fosite/pull/466 and then this PR will also pass.

**Comment by aeneasr** [comment]
Ah sorry, I forgot that this failure was on master :)

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
Awesome, thank you for your contribution!

**Pull request merged**